samples/matter/common/src/bridge/ble_connectivity_manager.h

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#pragma once
+
+#include "ble_bridged_device.h"
+#include "bridged_device_data_provider.h"
+
+#include <zephyr/bluetooth/addr.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/kernel.h>
+
+class BLEConnectivityManager {
+public:
+	struct ScannedDevice {
+		bt_addr_le_t mAddr;
+		bt_le_conn_param mConnParam;
+	};
+
+	static constexpr uint16_t kScanTimeoutMs = 10000;
+	static constexpr uint16_t kMaxScannedDevices = 16;
+	/* One BT connection is reserved for the Matter service purposes. */
+	static constexpr uint16_t kMaxCreatedDevices = CONFIG_BT_MAX_CONN - 1;
+	static constexpr uint8_t kMaxServiceUuids = CONFIG_BT_SCAN_UUID_CNT;
+
+	CHIP_ERROR Init(bt_uuid **serviceUuids, uint8_t serviceUuidsCount);
+	CHIP_ERROR Scan();
+	CHIP_ERROR StopScan();
+	CHIP_ERROR Connect(uint8_t index, BLEBridgedDevice::DeviceConnectedCallback callback, void *context,
+			   bt_uuid *serviceUuid);
+	BLEBridgedDevice *FindBLEBridgedDevice(struct bt_conn *conn);
+
+	static void FilterMatch(struct bt_scan_device_info *device_info, struct bt_scan_filter_match *filter_match,
+				bool connectable);
+	static void ScanTimeoutCallback(k_timer *timer);
+	static void ScanTimeoutHandle(intptr_t context);
+	static void ConnectionHandler(struct bt_conn *conn, uint8_t conn_err);
+	static void DisconnectionHandler(struct bt_conn *conn, uint8_t reason);
+	static void DiscoveryCompletedHandler(struct bt_gatt_dm *dm, void *context);
+	static void DiscoveryNotFound(struct bt_conn *conn, void *context);
+	static void DiscoveryError(struct bt_conn *conn, int err, void *context);
+
+private:
+	bool mScanActive;
+	k_timer mScanTimer;
+	uint8_t mScannedDevicesCounter;
+	uint8_t mCreatedDevicesCounter;
+	ScannedDevice mScannedDevices[kMaxScannedDevices];
+	BLEBridgedDevice mCreatedDevices[kMaxCreatedDevices];
+	bt_uuid *mServicesUuid[kMaxServiceUuids];
+	uint8_t mServicesUuidCount;
+
+	friend BLEConnectivityManager &GetBLEConnectivityManager();
+	static BLEConnectivityManager sBLEConnectivityManager;
+};
+
+inline BLEConnectivityManager &GetBLEConnectivityManager()
```

Comment:
This is my subjective opinion so I leave the decision up to you @kkasperczyk-no but I prefer the singleton implementation with local static instance being returned by reference for the following reasons:
- less code
- more encapsulation
- might be less prone to the static initialization fiasco
- no need to additionally define the static object in the cpp file
- no need for friendship relationship

Also note that to achieve "the real" singleton pattern, we would need to make the constructor of the `BLEConnectivityManager` private.

Reply:
Do you mean something like:     

> static BLEConnectivityManager & Instance()
>     {
>         static BLEConnectivityManager sInstance;
>         return sInstance;
>     }

?

Reply:
Yes.

Reply:
@markaj-nordic I refactored `BLEConnectivityManager` and `BridgeManager` to follow design recommended by you.